core/src/main/java/org/apache/iceberg/taskio/BaseTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.taskio;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DataFiles;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Tasks;
+
+abstract class BaseTaskWriter<T> implements TaskWriter<T> {
+  protected static final int ROWS_DIVISOR = 1000;
+
+  private final List<DataFile> completedFiles = Lists.newArrayList();
+  private final PartitionSpec spec;
+  private final FileFormat format;
+  private final FileAppenderFactory<T> appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+
+  protected BaseTaskWriter(PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
+                           OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+    this.spec = spec;
+    this.format = format;
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+  }
+
+  @Override
+  public void abort() throws IOException {
+    close();
+
+    // clean up files created by this writer
+    Tasks.foreach(completedFiles)
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(file -> io.deleteFile(file.path().toString()));
+  }
+
+  @Override
+  public List<DataFile> pollCompleteFiles() {
+    if (completedFiles.size() > 0) {
+      List<DataFile> dataFiles = ImmutableList.copyOf(completedFiles);
+      completedFiles.clear();
+      return dataFiles;
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  protected OutputFileFactory outputFileFactory() {
+    return this.fileFactory;
+  }
+
+  WrappedFileAppender createWrappedFileAppender(PartitionKey partitionKey,
```

Comment:
If this is a refactor, I'd prefer to keep the existing structure the way it was, with `openCurent`, `closeCurrent`, and `writeInternal`. That way we aren't introducing additional changes in this PR. If we want to refactor how writers work, we can do that separately.

Reply:
If there is concern about `openCurrent()` having code specific to sub-classes, then we can change it to accept an `EncryptedOutputFile` when it is called. Then `BaseWriter` wouldn't need to have any logic other than for releasing files that get too large. That would probably require a factory method for `EncryptedOutputFile` since it is called in `writeInternal` though. Probably easier just to leave this as it is for now to make this easier to review and get in.

Reply:
Okay, looks like this is intended to support the fan-out writer because the current `BaseWriter` only supports one open file at a time. I see why there are more changes.